packages/iov-bns/src/decode.ts

```diff
@@ -192,6 +215,132 @@ export function decodeParticipants(
+
+export function decodeElectionRule(
+  prefix: "iov" | "tiov",
+  rule: codecImpl.gov.IElectionRule & Keyed,
+): ElectionRule {
+  const { id } = decodeVersionedId(codecImpl.orm.VersionedIDRef.decode(rule._id));
+  const electorateId = new BN(ensure(rule.electorateId, "electorateId"));
+  return {
+    id: id,
+    version: asIntegerNumber(ensure(rule.version, "version")),
+    admin: encodeBnsAddress(prefix, ensure(rule.admin, "admin")),
+    electorateId: electorateId.toNumber(),
+    title: ensure(rule.title, "title"), // must not be an empty string
+    votingPeriod: asIntegerNumber(ensure(rule.votingPeriod, "votingPeriod")),
+    threshold: decodeFraction(ensure(rule.threshold, "threshold")),
+    quorum: rule.quorum ? decodeFraction(rule.quorum) : null,
+  };
+}
+
+function decodeProposalExecutorResult(result: codecImpl.gov.Proposal.ExecutorResult): ProposalExecutorResult {
+  switch (result) {
+    case codecImpl.gov.Proposal.ExecutorResult.PROPOSAL_EXECUTOR_RESULT_INVALID:
+      throw new Error("PROPOSAL_EXECUTOR_RESULT_INVALID is not allowed");
+    case codecImpl.gov.Proposal.ExecutorResult.PROPOSAL_EXECUTOR_RESULT_NOT_RUN:
+      return ProposalExecutorResult.NotRun;
+    case codecImpl.gov.Proposal.ExecutorResult.PROPOSAL_EXECUTOR_RESULT_SUCCESS:
+      return ProposalExecutorResult.Succeeded;
+    case codecImpl.gov.Proposal.ExecutorResult.PROPOSAL_EXECUTOR_RESULT_FAILURE:
+      return ProposalExecutorResult.Failed;
+    default:
+      throw new Error("Received unknown value for proposal executor result");
+  }
+}
+
+function decodeProposalResult(result: codecImpl.gov.Proposal.Result): ProposalResult {
+  switch (result) {
+    case codecImpl.gov.Proposal.Result.PROPOSAL_RESULT_INVALID:
+      throw new Error("PROPOSAL_RESULT_INVALID is not allowed");
+    case codecImpl.gov.Proposal.Result.PROPOSAL_RESULT_UNDEFINED:
+      return ProposalResult.Undefined;
+    case codecImpl.gov.Proposal.Result.PROPOSAL_RESULT_ACCEPTED:
+      return ProposalResult.Accepted;
+    case codecImpl.gov.Proposal.Result.PROPOSAL_RESULT_REJECTED:
+      return ProposalResult.Rejected;
+    default:
+      throw new Error("Received unknown value for proposal result");
+  }
+}
+
+function decodeProposalStatus(status: codecImpl.gov.Proposal.Status): ProposalStatus {
+  switch (status) {
+    case codecImpl.gov.Proposal.Status.PROPOSAL_STATUS_INVALID:
+      throw new Error("PROPOSAL_STATUS_INVALID is not allowed");
+    case codecImpl.gov.Proposal.Status.PROPOSAL_STATUS_SUBMITTED:
+      return ProposalStatus.Submitted;
+    case codecImpl.gov.Proposal.Status.PROPOSAL_STATUS_CLOSED:
+      return ProposalStatus.Closed;
+    case codecImpl.gov.Proposal.Status.PROPOSAL_STATUS_WITHDRAWN:
+      return ProposalStatus.Withdrawn;
+    default:
+      throw new Error("Received unknown value for proposal status");
+  }
+}
+
+function decodeRawProposalOption(rawOption: Uint8Array): ProposalOption {
+  const po = codecImpl.app.ProposalOptions.decode(rawOption);
```

Comment:
Prefer verbose names

Reply:
Right. Before I extracted this into a separate method, all other obvious names were taken. But now we can adjust easily.